vm/actor/src/builtin/miner/bitfield_queue.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::{actor_error, ActorError, ExitCode};
+
+use super::QuantSpec;
+use bitfield::BitField;
+use cid::Cid;
+use clock::ChainEpoch;
+use ipld_amt::{Amt, Error as AmtError};
+use ipld_blockstore::BlockStore;
+use std::collections::HashMap;
+use std::error::Error as StdError;
+
+/// Wrapper for working with an AMT[ChainEpoch]*Bitfield functioning as a queue, bucketed by epoch.
+/// Keys in the queue are quantized (upwards), modulo some offset, to reduce the cardinality of keys.
+pub struct BitFieldQueue<'db, BS> {
+    pub amt: Amt<'db, BitField, BS>,
+    quant: QuantSpec,
+}
+
+impl<'db, BS: BlockStore> BitFieldQueue<'db, BS> {
+    pub fn new(store: &'db BS, root: &Cid, quant: QuantSpec) -> Result<Self, AmtError> {
+        Ok(Self {
+            amt: Amt::load(root, store)?,
+            quant,
+        })
+    }
+
+    /// Adds values to the queue entry for an epoch.
+    pub fn add_to_queue(
+        &mut self,
+        raw_epoch: ChainEpoch,
+        values: &BitField,
+    ) -> Result<(), Box<dyn StdError>> {
+        if values.is_empty() {
+            // nothing to do.
+            return Ok(());
+        }
+
+        let epoch = self.quant.quantize_up(raw_epoch);
+
+        let bitfield = self
+            .amt
+            .get(epoch as u64)
+            .map_err(|e| format!("failed to lookup queue epoch {}: {:?}", epoch, e))?
+            .ok_or_else(|| actor_error!(ErrNotFound; ""))?;
+
+        self.amt
+            .set(epoch as u64, &bitfield | values)
+            .map_err(|e| format!("failed to set queue epoch {}: {:?}", epoch, e))?;
+
+        Ok(())
+    }
+
+    pub fn add_to_queue_values(
+        &mut self,
+        epoch: ChainEpoch,
+        values: &[u64],
+    ) -> Result<(), Box<dyn StdError>> {
+        self.add_to_queue(epoch, &values.iter().map(|&i| i as usize).collect())
```

Comment:
Yeah I assume that's what my reasoning was, creating a bitfield from an empty iterator should be completely allocation-free because `Vec` and `HashMap` don't allocate when they're empty. I'll add the check though, might as well match their code when it behaves the same way

Reply:
Just also future proofs it in case something else changes internally